osquery/tables/system/linux/pci_devices.cpp

```diff
@@ -26,38 +31,191 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
 const std::string kPCIKeyID = "PCI_ID";
 const std::string kPCIKeyDriver = "DRIVER";
+const std::string kPCISubsysID = "PCI_SUBSYS_ID";
+
+const std::string kPciIdsPath = "/usr/share/misc/pci.ids";
+
+PciDB::PciDB(std::istream& db_filestream) {
+  // pci.ids keep track of subsystem information of vendor and models
+  // sequentially so we keep track of what the current vendor and models are.
+  std::string cur_vendor, cur_model, line;
+  while (std::getline(db_filestream, line)) {
+    line = line.substr(0, line.find_first_of('#'));
+    boost::trim_right(line);
+    if (line.size() < 7) {
+      continue;
+    }
+
+    switch (line.find_first_of("0123456789abcdef")) {
+    case 0: {
+      // Vendor info.
+      cur_vendor = line.substr(0, 4);
+
+      // Once we get to illege vendor section we can stop since we're not
+      // currently parsing device classes.
+      if (cur_vendor == "ffff") {
+        return;
+      }
+
+      // Setup current vendor.
+      auto& vendor = db_[cur_vendor];
+      vendor.id = cur_vendor;
+      // Bump 2 chars to account for whitespace separation..
+      vendor.name = line.substr(6);
+      db_[cur_vendor] = vendor;
+
+      break;
+    }
+
+    case 1: {
+      // Model info.
+      auto vendor = db_.find(cur_vendor);
+      if (vendor != db_.end() && line.size() > 7) {
+        cur_model = line.substr(1, 4);
+
+        // Set up current model under the current vendor.
+        auto& model = db_[cur_vendor].models[cur_model];
+        model.id = cur_model;
+        // Bump 2 chars to account for whitespace separation.
+        model.desc = line.substr(7);
+        vendor->second.models[cur_model] = model;
+
+      } else {
+        VLOG(1) << "Unexpected error while parsing pci.ids: current vendor ID "
+                << cur_vendor << " does not exist in DB yet";
+      }
+
+      break;
+    }
+
+    case 2: {
+      // Subsystem info;
+      auto vendor = db_.find(cur_vendor);
+      if (vendor == db_.end()) {
+        VLOG(1) << "Unexpected error while parsing pci.ids: current vendor ID "
+                << cur_vendor << "does not exist in DB yet";
+        continue;
+      }
+
+      auto model = vendor->second.models.find(cur_model);
+      if (model == vendor->second.models.end()) {
+        VLOG(1) << "Unexpected error while parsing pci.ids: current model ID "
+                << cur_model << "does not exist in DB yet";
+        continue;
+      }
+
+      if (line.size() > 11) {
+        // Store current subsystem information under current vendor and model.
+        auto subsystemInfo = line.substr(11);
+        boost::trim(subsystemInfo);
+        model->second.subsystemInfo[line.substr(2, 9)] = subsystemInfo;
+      }
+
+      break;
+    }
+
+    default:
+      VLOG(1) << "Unexpected pci.ids line format";
+    }
+  }
+}
+
+Status PciDB::getVendorName(const std::string& vendor_id, std::string& name) {
+  return vendor(vendor_id, [&](const PciVendor& v) {
```

Comment:
I would kindly recommend getting rid of the vendor function(here and in other places). To me, It's okay to repeat two lines of code if it helps code readability.